Add bsdtar support to swupd

rojkov opened the pull request.

Commit:
recipes-swupd: Remove swupd recipes from meta-ostro

These recipes are supposed to be a part of the meta-swupd
layer that cn be used not only in Ostro.

Signed-off-by: Dmitry Rozhkov <dmitry.rozhkov@linux.intel.com>

Commit:
swupd-client: Update the client and server to use bsdtar

This patch adds bsdtar support to swupd-client and swupd-server
and enables it.

The reason why it's done this way is that:
    - bsdtar works better with IMA (opens files only once and then
      updates content and xattrs together);
    - swupd remains fully functional, including xattrs support,
      even when a distro disables GPLv3 licensed code.

Signed-off-by: Dmitry Rozhkov <dmitry.rozhkov@linux.intel.com>

Commit:
libarchive: Enable xz and xattrs in native builds

Some recipes may need to generate archives during build time
that are consumed on target devices, e. g. software updates.

Since the format of those archives must be identical configure
the native tool the same way as it's done for target builds:
enable xattrs if it's a distro feature. Also make the xz
compression type enabled always in addition to zlib and bzip2.

Signed-off-by: Dmitry Rozhkov <dmitry.rozhkov@linux.intel.com>

Commit:
swupd-client: Add ostro specific tweaks to swupd

Ostro uses an update mechanism which differs from what's
used in ClearLinux. This patch adds Ostro specific EFI updater
to the swupd client.

Signed-off-by: Dmitry Rozhkov <dmitry.rozhkov@linux.intel.com>

swupd-client: add rdep on gptfdisk to satisfy EFI updater

Ostro's swupd uses efi_combo_updater app to update the
boot partitions. This apps relies on sgdisk presence which
is a part of gptfdisk.

Signed-off-by: Dmitry Rozhkov <dmitry.rozhkov@linux.intel.com>

Commit:
swupd-client: Tweak os-release to conform ClearLinux format

swupd-client checks VERSION_ID, which must match the OS_VERSION
used for generating swupd bundles in the current build.

Also move os-release from /etc to /usr/lib because
swupd-client doesn't update files in /etc.

Signed-off-by: Dmitry Rozhkov <dmitry.rozhkov@linux.intel.com>

pohly commented on meta-swupd/recipes-core/swupd-client/os-release.bbappend:
```diff
@@ -0,0 +1,13 @@
+
+# As swupd-client doesn't update files in /etc move os-release's
+# content to /usr/lib
+do_install_append() {
```
I'm okay with keeping the do_install_append based on the explanation.

One could also have argued that this should be handled instead by the "stateless" work which is on-going, but for now, let's keep this workaround.

pohly commented on meta-ostro/recipes-swupd/swupd-client/swupd-client/efi_combo_updater.c:
```diff
@@ -0,0 +1,136 @@
+#include <stdio.h>
```
This whole file should be rewritten, but that's something for a different PR. For now, let's use it as it is.

The reason for rewriting:
- was meant to remove a dependency on /bin/sh, but now still uses popen() and thus still depends on /bin/sh
- assert() for error checking

incandescant commented on meta-swupd/recipes-core/swupd-server/swupd-server_2.53.bb:
```diff
@@ -30,6 +33,6 @@ do_install_append () {
 }
 
-RDEPENDS_${PN} = "tar rsync"
+RDEPENDS_${PN} = "libarchive rsync"
```
Shouldn't this be bsdtar? That's what is added to the RDEPENDS for swupd-client above.

rojkov replied:
When replaced with bsdtar it breaks with

```
Parsing recipes: 100% |###############################################################################| Time: 00:00:09
Parsing of 1981 .bb files complete (0 cached, 1981 parsed). 2586 targets, 1279 skipped, 0 masked, 0 errors.
NOTE: Resolving any missing task queue dependencies
ERROR: Nothing RPROVIDES 'bsdtar-native' (but virtual:native:/home/rojkov/work/ostro-os/meta-swupd/recipes-core/swupd-
server/swupd-server_2.53.bb RDEPENDS on or otherwise requires it)                                                    
NOTE: Runtime target 'bsdtar-native' is unbuildable, removing...
Missing or unbuildable dependency chain was: ['bsdtar-native']
ERROR: Required build target 'ostro-image-reference' has no buildable providers.
Missing or unbuildable dependency chain was: ['ostro-image-reference', 'swupd-server-native', 'bsdtar-native']
```

pohly replied:
That's the native.bbclass issue we talked about earlier. libarchive has a "bsdtar" PACKAGE, but that doesn't seem to lead to a corresponding RPROVIDES in libarchive-native.

pohly replied:
However, the workaround in swupd-server probably should be:

```
# Work around lack of "RPROVIDES = bsdtar-native" in libarchive-native.
RDEPENDS_${PN}_class-target = "bsdtar rsync"
RDEPENDS_${PN}_class-native = "libarchive rsync"
```

rojkov replied:
Yep, that adds clarity. I'll update the commit

rojkov commented:
Updated swupd-server_2.53.bb with separate RDEPENDS for target and native classes. Force pushed.

rojkov commented:
Added one more commit introducing runtime dependency on gptfdisk for Ostro's swupd-client.

Commit:
swupd-client: Update Upstream-Status for Fix-build-failure-on-Yocto.patch

The patch has been accepted upstream

Signed-off-by: Dmitry Rozhkov <dmitry.rozhkov@linux.intel.com>

Commit:
swupd-client: Update patch fixing quotes in os-release

Update 0001-Tolerate-quotes-in-os-release-files.patch to the
final version that got accepted upstream.

Signed-off-by: Dmitry Rozhkov <dmitry.rozhkov@linux.intel.com>

rojkov commented:
Dropped number prefixes from the patches that weren't introduced in this PR to avoid useless noise in diffs. Force pushed.

rojkov commented:
Fixed a typo in 3dbf800 commit message, force pushed

incandescant commented on meta-swupd/recipes-core/swupd-client/swupd-client/0001-Tolerate-quotes-in-os-release-files.patch:
```diff
@@ -10,19 +10,18 @@ to the format defined in systemd.
-Signed-off-by: Dmitry Rozhkov <dmitry.rozhkov@intel.com>
-
-Upstream-Status: Backport (v3.0.0+)
+Upstream-Status: Accepted
```
Why isn't backport right here? 

This patch was accepted upstream and we've backported it. By including the Backport tag and a version number we can more easily determine when this patch can be dropped.

http://www.openembedded.org/wiki/Commit_Patch_Message_Guidelines#Patch_Header_Recommendations

incandescant commented on meta-swupd/recipes-core/swupd-client/swupd-client/Fix-build-failure-on-Yocto.patch:
```diff
@@ -6,7 +6,7 @@ Subject: [PATCH] Fix build failure on Yocto
 _DATA twice. We can use EXTRA_DIST than dist_.
 
-Upstream-Status: Inappropriate
+Upstream-Status: Accepted
```
I think this counts as a backport also? See: http://www.openembedded.org/wiki/Commit_Patch_Message_Guidelines#Patch_Header_Recommendations

rojkov commented:
Fixed Upstream-Status tags in patches (Accepted -> Backported (v3.0.0+)). force pushed.

pohly commented:
If there are further changes expected or needed before including into meta-swupd, please comment on the commits in pohly/ostro-os:meta-swupd3.
